compiler/crates/resolution-path/src/test.rs

```diff
@@ -390,3 +413,1183 @@ fn fragment_argument_directive() {
+        );
+    })
+}
+
+#[test]
+fn union_type_definition_directive() {
+    let source = r#"
+        union Foo @bar = Baz
+        "#;
+    test_schema_resolution(source, "bar", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::ConstantDirectiveName(ConstantDirectivePath {
+                    inner: _,
+                    parent: ConstantDirectiveParent::UnionTypeDefinition(_),
+                })
+            })
+        );
+    })
+}
+
+// ## Union Type Extensions
+
+#[test]
+fn union_type_extension_name() {
+    let source = r#"
+        extend union Foo = Bar | Baz
+        "#;
+    test_schema_resolution(source, "Foo", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::UnionTypeExtensionName(_),
+            })
+        );
+    })
+}
+
+#[test]
+fn union_type_extension_member_name() {
+    let source = r#"
+        extend union Foo = Bar | Baz
+        "#;
+    test_schema_resolution(source, "Bar", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::UnionTypeMemberType(
+                    UnionTypeMemberParent::UnionTypeExtension(_),
+                )
+            })
+        );
+    })
+}
+
+#[test]
+fn union_type_extension_directive() {
+    let source = r#"
+        extend union Foo @bar = Baz
+        "#;
+    test_schema_resolution(source, "bar", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::ConstantDirectiveName(ConstantDirectivePath {
+                    inner: _,
+                    parent: ConstantDirectiveParent::UnionTypeExtension(_),
+                })
+            })
+        );
+    })
+}
+
+// ## Interface Types
+
+#[test]
+fn interface_type_definition_name() {
+    let source = r#"
+        interface Foo {
+            bar: Baz
+        }
+        "#;
+    test_schema_resolution(source, "Foo", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::InterfaceTypeDefinitionName(_),
+            })
+        );
+    })
+}
+
+#[test]
+fn interface_type_definition_implements_interface() {
+    let source = r#"
+        interface Foo implements Bar {
+            baz: Qux
+        }
+        "#;
+    test_schema_resolution(source, "Bar", |resolved| {
+        assert_matches!(
+            resolved,
+            ResolutionPath::Ident(IdentPath {
+                inner: _,
+                parent: IdentParent::ImplementedInterfaceName(
+                    ImplementedInterfaceParent::InterfaceTypeDefinition(_),
```

Comment:
Same problem here. An implemented interface identifier can appear on interface/object definition/extension. I didn't want to create 4 IdentParents, but this Parent in Parent also feels very weird 😅

Reply:
I think what you've done here feels like a reasonable tradeoff. I don't think any one looking at the resolution path will need to handle the individual cases here, so it shouldn't be a problem in practice. They'll just check "oh, this identifier's parent is an `ImplementedInterfaceName`, I'll interpret it as an interface name". They shouldn't need to care about the four different places that a ImplementedInterfaceName can appear, so can just ignore it.

At the same time, what you have here does preserve the full information, which feels good.

Is there some other tension this creates that I'm missing?

Reply:
No, it works for me. I was just worried since this would be the only place where we have a parent in a parent :D